stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -737,10 +743,12 @@ julia> rank(diagm(0 => [1, 0.001, 2]), atol=0.00001, rtol=0.00001)
 3
 ```
 """
-function rank(A::AbstractMatrix; atol=0.0, rtol=0.0)
+function rank(A::AbstractMatrix; atol::Real=0.0, rtol::Real=atol>0 ? (min(size(A)...)*eps(real(float(one(eltype(A)))))) : 0.0)
```

Comment:
This expression is not type-stable, since `0.0` is always `Float64` but the other branch is not. You could use
```jl
(min(size(A)...)*eps(real(float(one(eltype(A))))))*iszero(atol)
```
instead